### PR TITLE
Refactor services on the server

### LINKED ISSUE #2867: Artemis does not startup reliably
### Describe the bug
Artemis sometimes does not start up but exits with an exception.

<details closed>
<summary>Log</summary>
2021-02-13 13:04:45,427 INFO success: application entered RUNNING state, process has stayed up for > than 1 seconds (startsecs)

█████╗ ██████╗ ████████╗███████╗███╗ ███╗██╗███████╗

██╔══██╗██╔══██╗╚══██╔══╝██╔════╝████╗ ████║██║██╔════╝

███████║██████╔╝ ██║ █████╗ ██╔████╔██║██║███████╗

██╔══██║██╔══██╗ ██║ ██╔══╝ ██║╚██╔╝██║██║╚════██║

██║ ██║██║ ██║ ██║ ███████╗██║ ╚═╝ ██║██║███████║

╚═╝ ╚═╝╚═╝ ╚═╝ ╚═╝ ╚══════╝╚═╝ ╚═╝╚═╝╚══════╝

:: JHipster 🤓 :: Running Spring Boot 2.4.2 ::

:: https://www.jhipster.tech ::

2021-02-13 13:04:49.116 INFO 490 --- [ main] de.tum.in.www1.artemis.ArtemisApp : Starting ArtemisApp using Java 15.0.2 on d76d75f8a584 with PID 490 (/opt/Artemis/Artemis.jar started by artemis in /opt/Artemis)

2021-02-13 13:04:49.121 INFO 490 --- [ main] de.tum.in.www1.artemis.ArtemisApp : The following profiles are active: prod,jenkins,gitlab,artemis,scheduling

2021-02-13 13:05:11.496 INFO 490 --- [ main] d.t.i.www1.artemis.config.WebConfigurer : Web application configuration, using profiles: prod

2021-02-13 13:05:11.499 INFO 490 --- [ main] d.t.i.www1.artemis.config.WebConfigurer : Web application fully configured

2021-02-13 13:05:12.102 INFO 490 --- [ main] d.t.i.w.a.config.LiquibaseConfiguration : Liquibase is enabled

2021-02-13 13:05:14.621 WARN 490 --- [ main] d.t.i.w.a.config.CacheConfiguration : No discovery service is set up, Hazelcast cannot create a cluster.

2021-02-13 13:05:14.621 INFO 490 --- [ main] d.t.i.w.a.config.CacheConfiguration : Binding Hazelcast to interface 127.0.0.1

2021-02-13 13:05:15.956 WARN 490 --- [ main] com.hazelcast.cp.CPSubsystem : [127.0.0.1]:5701 [dev] [4.1.1] CP Subsystem is not enabled. CP data structures will operate in UNSAFE mode! Please note that UNSAFE mode will not provide strong consistency guarantees.

2021-02-13 13:05:26.988 INFO 490 --- [ main] d.t.i.w.a.service.connectors.GitService : file.encoding=UTF-8

2021-02-13 13:05:26.988 INFO 490 --- [ main] d.t.i.w.a.service.connectors.GitService : sun.jnu.encoding=ANSI_X3.4-1968

2021-02-13 13:05:26.989 INFO 490 --- [ main] d.t.i.w.a.service.connectors.GitService : Default Charset=UTF-8

2021-02-13 13:05:26.989 INFO 490 --- [ main] d.t.i.w.a.service.connectors.GitService : Default Charset in Use=UTF8

2021-02-13 13:05:26.991 INFO 490 --- [ main] d.t.i.w.a.service.connectors.GitService : GitService will use ssh keys as authentication method to interact with remote git repositories

2021-02-13 13:05:28.467 INFO 490 --- [ main] d.t.i.w.a.c.w.WebsocketConfiguration : Did NOT enable StompBrokerRelay for WebSocket messages

2021-02-13 13:05:29.740 INFO 490 --- [ main] d.t.i.w.a.config.SentryConfiguration : Sentry is disabled: Provide a DSN to enable Sentry.

2021-02-13 13:05:31.619 WARN 490 --- [ main] ConfigServletWebServerApplicationContext : Exception encountered during context initialization - cancelling refresh attempt: org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'assessmentDashboardService' defined in URL [jar:file:/opt/Artemis/Artemis.jar!/BOOT-INF/classes!/de/tum/in/www1/artemis/service/AssessmentDashboardService.class]: Unsatisfied dependency expressed through constructor parameter 2; nested exception is org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'submissionService': Unsatisfied dependency expressed through method 'setCourseService' parameter 0; nested exception is org.springframework.beans.factory.BeanCurrentlyInCreationException: Error creating bean with name 'courseService': Bean with name 'courseService' has been injected into other beans [examService] in its raw version as part of a circular reference, but has eventually been wrapped. This means that said other beans do not use the final version of the bean. This is often the result of over-eager type matching - consider using 'getBeanNamesForType' with the 'allowEagerInit' flag turned off, for example.

2021-02-13 13:05:31.643 WARN 490 --- [ main] c.h.h.i.IHazelcastInstanceFactory : hibernate.cache.hazelcast.shutdown_on_session_factory_close property is set to 'false'. Leaving current HazelcastInstance active! (Warning: Do not disable Hazelcast hazelcast.shutdownhook.enabled property!)

2021-02-13 13:05:31.664 INFO 490 --- [ main] d.t.i.w.a.config.CacheConfiguration : Closing Cache Manager

2021-02-13 13:05:31.677 WARN 490 --- [ main] o.a.c.loader.WebappClassLoaderBase : The web application [ROOT] appears to have started a thread named [SentryHostnameCache-0] but has failed to stop it. This is very likely to create a memory leak. Stack trace of thread:

java.base@15.0.2/jdk.internal.misc.Unsafe.park(Native Method)

java.base@15.0.2/java.util.concurrent.locks.LockSupport.park(LockSupport.java:341)

java.base@15.0.2/java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionNode.block(AbstractQueuedSynchronizer.java:505)

java.base@15.0.2/java.util.concurrent.ForkJoinPool.managedBlock(ForkJoinPool.java:3137)

java.base@15.0.2/java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:1614)

java.base@15.0.2/java.util.concurrent.LinkedBlockingQueue.take(LinkedBlockingQueue.java:435)

java.base@15.0.2/java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1056)

java.base@15.0.2/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1116)

java.base@15.0.2/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)

java.base@15.0.2/java.lang.Thread.run(Thread.java:832)

2021-02-13 13:05:31.788 ERROR 490 --- [ main] o.s.boot.SpringApplication : Application run failed

org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'assessmentDashboardService' defined in URL [jar:file:/opt/Artemis/Artemis.jar!/BOOT-INF/classes!/de/tum/in/www1/artemis/service/AssessmentDashboardService.class]: Unsatisfied dependency expressed through constructor parameter 2; nested exception is org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'submissionService': Unsatisfied dependency expressed through method 'setCourseService' parameter 0; nested exception is org.springframework.beans.factory.BeanCurrentlyInCreationException: Error creating bean with name 'courseService': Bean with name 'courseService' has been injected into other beans [examService] in its raw version as part of a circular reference, but has eventually been wrapped. This means that said other beans do not use the final version of the bean. This is often the result of over-eager type matching - consider using 'getBeanNamesForType' with the 'allowEagerInit' flag turned off, for example.

at org.springframework.beans.factory.support.ConstructorResolver.createArgumentArray(ConstructorResolver.java:800)

at org.springframework.beans.factory.support.ConstructorResolver.autowireConstructor(ConstructorResolver.java:229)

at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.autowireConstructor(AbstractAutowireCapableBeanFactory.java:1356)

at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1206)

at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:571)

at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:531)

at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:335)

at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:234)

at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:333)

at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:208)

at org.springframework.beans.factory.support.DefaultListableBeanFactory.preInstantiateSingletons(DefaultListableBeanFactory.java:944)

at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:923)

at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:588)

at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.refresh(ServletWebServerApplicationContext.java:144)

at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:767)

at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:759)

at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:426)

at org.springframework.boot.SpringApplication.run(SpringApplication.java:326)

at de.tum.in.www1.artemis.ArtemisApp.main(ArtemisApp.java:60)

at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)

at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:64)

at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)

at java.base/java.lang.reflect.Method.invoke(Method.java:564)

at org.springframework.boot.loader.MainMethodRunner.run(MainMethodRunner.java:49)

at org.springframework.boot.loader.Launcher.launch(Launcher.java:107)

at org.springframework.boot.loader.Launcher.launch(Launcher.java:58)

at org.springframework.boot.loader.JarLauncher.main(JarLauncher.java:88)

Caused by: org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'submissionService': Unsatisfied dependency expressed through method 'setCourseService' parameter 0; nested exception is org.springframework.beans.factory.BeanCurrentlyInCreationException: Error creating bean with name 'courseService': Bean with name 'courseService' has been injected into other beans [examService] in its raw version as part of a circular reference, but has eventually been wrapped. This means that said other beans do not use the final version of the bean. This is often the result of over-eager type matching - consider using 'getBeanNamesForType' with the 'allowEagerInit' flag turned off, for example.

at org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor$AutowiredMethodElement.inject(AutowiredAnnotationBeanPostProcessor.java:722)

at org.springframework.beans.factory.annotation.InjectionMetadata.inject(InjectionMetadata.java:119)

at org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor.postProcessProperties(AutowiredAnnotationBeanPostProcessor.java:399)

at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.populateBean(AbstractAutowireCapableBeanFactory.java:1415)

at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:608)

at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:531)

at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:335)

at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:234)

at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:333)

at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:208)

at org.springframework.beans.factory.config.DependencyDescriptor.resolveCandidate(DependencyDescriptor.java:276)

at org.springframework.beans.factory.support.DefaultListableBeanFactory.doResolveDependency(DefaultListableBeanFactory.java:1380)

at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveDependency(DefaultListableBeanFactory.java:1300)

at org.springframework.beans.factory.support.ConstructorResolver.resolveAutowiredArgument(ConstructorResolver.java:887)

at org.springframework.beans.factory.support.ConstructorResolver.createArgumentArray(ConstructorResolver.java:791)

... 26 common frames omitted

Caused by: org.springframework.beans.factory.BeanCurrentlyInCreationException: Error creating bean with name 'courseService': Bean with name 'courseService' has been injected into other beans [examService] in its raw version as part of a circular reference, but has eventually been wrapped. This means that said other beans do not use the final version of the bean. This is often the result of over-eager type matching - consider using 'getBeanNamesForType' with the 'allowEagerInit' flag turned off, for example.

at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:638)

at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:531)

at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:335)

at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:234)

at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:333)

at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:208)

at org.springframework.beans.factory.config.DependencyDescriptor.resolveCandidate(DependencyDescriptor.java:276)

at org.springframework.beans.factory.support.DefaultListableBeanFactory.doResolveDependency(DefaultListableBeanFactory.java:1380)

at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveDependency(DefaultListableBeanFactory.java:1300)

at org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor$AutowiredMethodElement.inject(AutowiredAnnotationBeanPostProcessor.java:714)

... 40 common frames omitted

2021-02-13 13:05:31,809 INFO exited: application (exit status 1; not expected)
</details>

#### To Reproduce
1. Start Artemis
2. Repeat until Artemis does not start

#### Expected behavior
Artemis should start up. In my case outputing:

<details closed>
<summary>Log</summary>
█████╗ ██████╗ ████████╗███████╗███╗ ███╗██╗███████╗

██╔══██╗██╔══██╗╚══██╔══╝██╔════╝████╗ ████║██║██╔════╝

███████║██████╔╝ ██║ █████╗ ██╔████╔██║██║███████╗

██╔══██║██╔══██╗ ██║ ██╔══╝ ██║╚██╔╝██║██║╚════██║

██║ ██║██║ ██║ ██║ ███████╗██║ ╚═╝ ██║██║███████║

╚═╝ ╚═╝╚═╝ ╚═╝ ╚═╝ ╚══════╝╚═╝ ╚═╝╚═╝╚══════╝

:: JHipster 🤓 :: Running Spring Boot 2.4.2 ::

:: https://www.jhipster.tech ::

2021-02-13 13:05:37.607 INFO 713 --- [ main] de.tum.in.www1.artemis.ArtemisApp : Starting ArtemisApp using Java 15.0.2 on d76d75f8a584 with PID 713 (/opt/Artemis/Artemis.jar started by artemis in /opt/Artemis)

2021-02-13 13:05:37.613 INFO 713 --- [ main] de.tum.in.www1.artemis.ArtemisApp : The following profiles are active: prod,jenkins,gitlab,artemis,scheduling

2021-02-13 13:06:00.365 INFO 713 --- [ main] d.t.i.www1.artemis.config.WebConfigurer : Web application configuration, using profiles: prod

2021-02-13 13:06:00.368 INFO 713 --- [ main] d.t.i.www1.artemis.config.WebConfigurer : Web application fully configured

2021-02-13 13:06:01.003 INFO 713 --- [ main] d.t.i.w.a.config.LiquibaseConfiguration : Liquibase is enabled

2021-02-13 13:06:03.492 WARN 713 --- [ main] d.t.i.w.a.config.CacheConfiguration : No discovery service is set up, Hazelcast cannot create a cluster.

2021-02-13 13:06:03.493 INFO 713 --- [ main] d.t.i.w.a.config.CacheConfiguration : Binding Hazelcast to interface 127.0.0.1

2021-02-13 13:06:04.826 WARN 713 --- [ main] com.hazelcast.cp.CPSubsystem : [127.0.0.1]:5701 [dev] [4.1.1] CP Subsystem is not enabled. CP data structures will operate in UNSAFE mode! Please note that UNSAFE mode will not provide strong consistency guarantees.

2021-02-13 13:06:16.042 INFO 713 --- [ main] d.t.i.w.a.service.connectors.GitService : file.encoding=UTF-8

2021-02-13 13:06:16.042 INFO 713 --- [ main] d.t.i.w.a.service.connectors.GitService : sun.jnu.encoding=ANSI_X3.4-1968

2021-02-13 13:06:16.043 INFO 713 --- [ main] d.t.i.w.a.service.connectors.GitService : Default Charset=UTF-8

2021-02-13 13:06:16.043 INFO 713 --- [ main] d.t.i.w.a.service.connectors.GitService : Default Charset in Use=UTF8

2021-02-13 13:06:16.046 INFO 713 --- [ main] d.t.i.w.a.service.connectors.GitService : GitService will use ssh keys as authentication method to interact with remote git repositories

2021-02-13 13:06:17.518 INFO 713 --- [ main] d.t.i.w.a.c.w.WebsocketConfiguration : Did NOT enable StompBrokerRelay for WebSocket messages

2021-02-13 13:06:18.755 INFO 713 --- [ main] d.t.i.w.a.config.SentryConfiguration : Sentry is disabled: Provide a DSN to enable Sentry.

2021-02-13 13:06:21.458 INFO 713 --- [ main] a.s.s.ProgrammingExerciseScheduleService : Scheduled building the student submissions for 0 programming exercises with a buildAndTestAfterDueDate.

2021-02-13 13:06:27.310 INFO 713 --- [ main] de.tum.in.www1.artemis.ArtemisApp : Started ArtemisApp in 52.435 seconds (JVM running for 54.483)

2021-02-13 13:06:27.357 INFO 713 --- [ main] d.t.i.w.a.s.s.quiz.QuizScheduleService : QuizScheduleService was started to run repeatedly with 5.0 second delay.

2021-02-13 13:06:27.384 INFO 713 --- [ main] d.t.i.w.a.s.s.quiz.QuizScheduleService : Found 0 quiz exercises with planned start in the future

2021-02-13 13:06:27.468 INFO 713 --- [ main] d.t.in.www1.artemis.service.UserService : Update internal admin user artemis_admin

2021-02-13 13:06:29.145 INFO 713 --- [ main] de.tum.in.www1.artemis.ArtemisApp :

----------------------------------------------------------

Application 'Artemis' is running! Access URLs:

Local: http://localhost:8080/

External: http://172.29.0.6:8080/

Profile(s): [prod, jenkins, gitlab, artemis, scheduling]

----------------------------------------------------------

2021-02-13 13:07:17.577 INFO 713 --- [MessageBroker-1] o.s.w.s.c.WebSocketMessageBrokerStats : WebSocketSession[0 current WS(0)-HttpStream(0)-HttpPoll(0), 0 total, 0 closed abnormally (0 connect failure, 0 send limit, 0 transport error)], stompSubProtocol[null], stompBrokerRelay[null], inboundChannel[pool size = 0, active threads = 0, queued tasks = 0, completed tasks = 0], outboundChannel[pool size = 0, active threads = 0, queued tasks = 0, completed tasks = 0], sockJsScheduler[pool size = 1, active threads = 1, queued tasks = 0, completed tasks = 0]
</details>

### Environment
 - OS: https://github.com/LDAP/docker-artemis-server
 - Version: Commit: 21c1df3261140fb8833413edbed0e901ec29a477
</pre></details>

### PR DESCRIPTION
**Pull request by krusche**
This PR starts with a refactoring of server services and their dependencies
Fixes #2867 

### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/server.html).
- [x] Server: I added `@PreAuthorize` and check the course groups for all new REST Calls (security)
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.

### Motivation and Context
We have too many (circular) dependencies between resources and services. We should instead follow the design principles `low coupling and high cohesion`

### Description
* Reduce potential circular dependencies (in particular between exam, exercises and course)
* Move simple service methods (e.g. `findByIdElseThrow`) into repositories as `default` methods (Example: before this PR almost every service and resource class had dependencies to `UserService`, now almost all only depend on the simple `UserRepository` for authentication checks)
* Extract smaller services with a dedicated purpose (e.g. `ExamDateService`)
* Move methods into different services in which they make more sense (e.g. in the exam services)
* I also moved some files into different folders (to make the code structure easier to understand, in case we have too many merge conflicts with other PRs, I could also revert this)

### Steps for Testing
Basically **all areas** are affected, so we should click through the most important use cases.
I tried to be careful so that no functionality is changed

### Test Coverage
no changes (might be slightly higher as I deleted some unused code)

### Screenshots
no changes

### CODE REVIEW

**Review by airelawaleria — APPROVED** [review]
tested with @JohannesWeiss @simonzilker @marvinchimeka. Tested all exercises, lectures  and exam mode

**Review by simonzilker — APPROVED** [review]
Tested in testing session (also exam mode). No issues found.

**Review by marvinchimeka — DISMISSED** [review]
Could not see any issues.

**Review by JohannesWeiss — DISMISSED** [review]
Tested in testing session as well, could not find any new problems.

**Review by valeriebucher — APPROVED** [review]
Works fine for me, I didn't find an issue!

**Review by DenizShabani — APPROVED** [review]
Looks good!

File: src/main/java/de/tum/in/www1/artemis/service/exam/StudentExamService.java [hunk 1]
```diff
@@ -22,6 +25,12 @@
 import de.tum.in.www1.artemis.domain.participation.StudentParticipation;
 import de.tum.in.www1.artemis.domain.quiz.*;
 import de.tum.in.www1.artemis.repository.*;
+import de.tum.in.www1.artemis.repository.UserRepository;
```
Comment (derLalla) [inline src/main/java/de/tum/in/www1/artemis/service/exam/StudentExamService.java:28]: there are some redundant imports of UserRepository in several Services.java

**Review by derLalla — CHANGES_REQUESTED** [review]
some redundant imports should be removed
